ykrt/src/thread_intercept.rs

```diff
@@ -1,11 +1,54 @@
 use std::os::raw::{c_int, c_void};
 
+use std::{ffi::CString, ptr::null_mut};
+
+use libc::{dlsym, free, malloc, pthread_attr_t, pthread_create, pthread_t};
+
+const SHADOW_STACK_SIZE: usize = 1000000;
+
+#[derive(Debug)]
+struct ThreadRoutine {
+    pub args: *mut c_void,
+    pub start_routine: extern "C" fn(*mut c_void) -> *mut c_void,
+}
+
+extern "C" fn wrap_thread_routine(arg: *mut c_void) -> *mut c_void {
+    let str = CString::new("shadowstack_0").unwrap();
+    unsafe {
```

Comment:
In general I'm nervous about big `unsafe` blocks like this because it lets us put lots of code in without thinking about whether it's safe or not. Ideally I prefer to use `unsafe` only on individual lines that *must* be unsafe.

Reply:
Ok. I can break it into separate, individual lines :)

Reply:
done here 👉 ced50ed5a4004fe77b616dedcf25c07bea088d4d